pkg/cmd/grafana-cli/commands/install_command.go

```diff
@@ -105,93 +129,103 @@ func InstallPlugin(pluginName, version string, c utils.CommandLine) error {
 	return err
 }
 
-func SelectVersion(plugin m.Plugin, version string) (m.Version, error) {
+func osAndArchString() string {
+	osString := strings.ToLower(runtime.GOOS)
+	arch := runtime.GOARCH
+	return osString + "-" + arch
+}
+
+func supportsCurrentArch(version *m.Version) bool {
+	if version.Arch == nil {
+		return true
+	}
+	for arch := range version.Arch {
+		if arch == osAndArchString() || arch == "any" {
+			return true
+		}
+	}
+	return false
+}
+
+func latestSupportedVersion(plugin *m.Plugin) *m.Version {
+	for _, ver := range plugin.Versions {
+		if supportsCurrentArch(&ver) {
+			return &ver
+		}
+	}
+	return nil
+}
+
+// SelectVersion returns latest version if none is specified or the specified version. If the version string is not
+// matched to existing version it errors out. It also errors out if version that is matched is not available for current
+// os and platform.
+func SelectVersion(plugin *m.Plugin, version string) (*m.Version, error) {
+	var ver *m.Version
 	if version == "" {
-		return plugin.Versions[0], nil
+		ver = &plugin.Versions[0]
 	}
 
 	for _, v := range plugin.Versions {
 		if v.Version == version {
-			return v, nil
+			ver = &v
 		}
 	}
 
-	return m.Version{}, errors.New("Could not find the version you're looking for")
+	if ver == nil {
+		return nil, xerrors.New("Could not find the version you're looking for")
+	}
+
+	latestForArch := latestSupportedVersion(plugin)
+	if latestForArch.Version == ver.Version {
+		return ver, nil
+	} else {
+		return nil, xerrors.Errorf("Version you want is not supported on your architecture. Latest suitable version is %v", latestForArch.Version)
+	}
 }
 
 func RemoveGitBuildFromName(pluginName, filename string) string {
 	r := regexp.MustCompile("^[a-zA-Z0-9_.-]*/")
 	return r.ReplaceAllString(filename, pluginName+"/")
 }
 
-var retryCount = 0
 var permissionsDeniedMessage = "Could not create %s. Permission denied. Make sure you have write access to plugindir"
 
-func downloadFile(pluginName, filePath, url string) (err error) {
-	defer func() {
-		if r := recover(); r != nil {
-			retryCount++
-			if retryCount < 3 {
-				fmt.Println("Failed downloading. Will retry once.")
-				err = downloadFile(pluginName, filePath, url)
-			} else {
-				failure := fmt.Sprintf("%v", r)
-				if failure == "runtime error: makeslice: len out of range" {
-					err = fmt.Errorf("Corrupt http response from source. Please try again")
-				} else {
-					panic(r)
-				}
-			}
-		}
-	}()
-
-	resp, err := http.Get(url) // #nosec
-	if err != nil {
```

Comment:
Bit of an issue in this old code. This does not check for response status. So even though for old versions this will return correctly 400 status because we do not know what is the arch of the grafana it will end up with:
```
zip: not a valid zip file
```
which is not helpful. Not sure how to make this better.

Reply:
@torkelo what is your opinion on this? Other solution could be not even return the arch specific plugin from the API if we do not sent arch from grafana (so old versions) in that case user would get plugin does not exists or something like that which does not seem much better and at the same time I think the API would be weirder and non standard (ie if you GET /plugins you would get just a subset and there wouldn't be a clear way to get all plugins no matter what arch, would need to add something like /plugins?all=true or something like that)

Reply:
Not sure I understand fully. So new grafana-cli adds arch as parameter, the change to grafana.com api is that backwards compatible for non backend plugins (it will need to be). ? 

So this is an error for old grafana versions trying to install a new grafana backend plugin? hard to modify that error message now I think.

Reply:
@torkelo 
Yes I think you get it but just to reiterate:
- `GET /plugins` will return all plugins with arch attribute and it is up to the client to filter them correctly
  - old grafana does not do anything so shows all plugins (so cli plugins list-remote will show everything without knowing which can be installed)
  - new grafana will show only available
- `GET /plugins/:id/downloads` has a new optional param `osAndArch`. If set it will return the zip file for specified arch, if not set it will return zip only if frontend plugin or if there is `any` arch zip file
  - old grafana can download non backend plugins as usual, for backend plugins it will error out with `zip: not a valid zip file` because it does not check for response status and we cannot know what arch the grafana is running on
  - new grafana sends the `osAndArch` so we pick the correct version (TODO I need to add the status check in the code) and does not allow you to install wrong plugin version with some sensible error message

Reply:
So adding `osAndArch` to `GET /plugins` in a way that without it, it will return only plugins which can be safely installed when we do not know the arch would be best backward compatible way, but not sure if that does not break other things that would expect it to return all plugins by default. It is also kind of nonintuitive behaviour.
Other solution would be to start versioning the grafana.com API on some level (either with some header or v2 prefix maybe) if we would consider that a breaking API change. That would make doing changes a bit easier in sense of keeping backward compatibility but probably harder for maintenance.

Reply:
So @DanCech pointed out to me [here](https://github.com/raintank/grafana-net/pull/790#issuecomment-503641061) that we are actually sending OS and ARCH already so it seems it would make sense to reuse those headers. Unfortunately for this issue it just complicates things a bit.
- We only send the headers in the `/plugin/:id` request that gets the metadata but not in the `/download` so we could know that plugin supports the grafana os/arch but than we still cannot select the correct version to download.
- In the `/plugin/:id` we won't be able to say if it is GF version before the change or after because the headers will be sent also from the old versions.

I thought we could just base it on raw grafana version that we also send in the headers but it seems like it is hardcoded to `master` https://github.com/grafana/grafana/commit/746257710b7adc3825cbe5aa7b21b76dade3440b#diff-1146cad1c61c159a07758ade40c6ac2aR12 and does not seem to change since (I assume we forgot to add it to release guidelines that it needs to be updated or something like that)

Reply:
That variable is set using a compile flag in build.go 

Not sure i understand the problem can’t we add these headers to the download call as well.  ?

Reply:
@torkelo 
> That variable is set using a compile flag in build.go
Ah ok then it can be used. 
> Not sure i understand the problem can’t we add these headers to the download call as well. ?
Yes I will add this the only issue is old versions of Grafana before the change how will they handle installing plugins and what would happen if they try to install binary plugin. As I mentioned at the moment they will return `zip: not a valid zip file` error which is not very helpful. With the grafana version sent, we can return some more helpful error message on the `/plugin/:id` call for the old Grafana versions. This would mean minimal changes to the API but still old Grafana will get `You need to upgrade your Grafana version to install this plugin` error or something like that.

Reply:
Ah I see. If you can find a way to show error to older Grafana versions that be great.

Maybe you can do that by checking the grafana dependency version plugins specify in plugin.json. This is a big missing thing we have yet to fix, that is return info error message when you try to install a plugin or plugin version that is not compatible with your Grafana version